packages/flame_riverpod/example/lib/main.dart

```diff
@@ -0,0 +1,158 @@
+        children: const [
+          Expanded(child: FlutterCountingComponent()),
+          Expanded(
+            child: RiverpodGameWidget.initializeWithGame(
+              uninitializedGame: RefExampleGame.new,
+            ),
+          )
+        ],
+      ),
+    );
+  }
+}
+
+class FlutterCountingComponent extends ConsumerWidget {
+  const FlutterCountingComponent({super.key});
+
+  @override
+  Widget build(BuildContext context, WidgetRef ref) {
+    final textStyle = Theme.of(context)
+        .textTheme
+        .headlineSmall
+        ?.copyWith(color: Colors.white);
+
+    final stream = ref.watch(countingStreamProvider);
+    return Material(
+      color: Colors.transparent,
+      child: Column(
+        children: [
+          Text('Flutter', style: textStyle),
+          stream.when(
+            data: (value) => Text('$value', style: textStyle),
+            error: (error, stackTrace) => Text('$error', style: textStyle),
+            loading: () => Text('Loading...', style: textStyle),
+          )
+        ],
+      ),
+    );
+  }
+}
+
+class RefExampleGame extends FlameGame with HasComponentRef {
+  RefExampleGame(WidgetRef ref) {
+    // Note: WidgetRef is not stored and is used to build a [ComponentRef],
+    // a wrapper around it.
+    HasComponentRef.widgetRef = ref;
+  }
+
+  @override
+  Future<void> onLoad() async {
+    await super.onLoad();
+    add(TextComponent(text: 'Flame'));
+    add(RiverpodAwareTextComponent());
+  }
+}
+
+class RiverpodGameWidget extends ConsumerStatefulWidget {
+  const RiverpodGameWidget.readFromProvider({super.key})
+      : uninitializedGame = null;
+  const RiverpodGameWidget.initializeWithGame({
+    super.key,
+    required this.uninitializedGame,
+  });
+
+  final FlameGame Function(WidgetRef ref)? uninitializedGame;
+
+  @override
+  ConsumerState<RiverpodGameWidget> createState() => _RiverpodGameWidgetState();
+}
+
+class _RiverpodGameWidgetState extends ConsumerState<RiverpodGameWidget> {
+  @override
+  void initState() {
+    WidgetsBinding.instance.addPostFrameCallback((timeStamp) {
+      if (widget.uninitializedGame is FlameGame Function(WidgetRef ref)) {
+        ref
+            .read(riverpodAwareGameProvider.notifier)
+            .set(widget.uninitializedGame!(ref));
+      }
+    });
+    super.initState();
+  }
+
+  @override
+  Widget build(BuildContext context) {
+    final game = ref.watch(riverpodAwareGameProvider);
+
+    if (game is! Game) {
+      return Container();
+    }
+
+    return GameWidget(game: game!);
+  }
+}
+
+class RiverpodAwareTextComponent extends PositionComponent
+    with HasComponentRef {
+  late TextComponent textComponent;
+  int currentValue = 0;
+
+  /// [onMount] should be used over [onLoad] to initialize subscriptions,
+  /// cancellation is handled for the user inside [onRemove],
+  /// which is only called if the [Component] was mounted.
+  @override
+  void onMount() {
+    super.onMount();
```

Comment:
Is there no equivalent to a "build" method in components?

We should be able to use ref.watch instead of ref.listen

What about:

```dart
class Example extends MyConsumerComponent {
  final component = TextComponent();

  @override
  void update(ComponentRef ref) {
    component.text = ref.watch(countingStreamProvider)?.value.toString() ?? '<loading>";
  }
}
```

Reply:
Flame works very differently from Flutter since it uses a game-loop, that's why it doesn't have a `build` method. It constantly renders and runs `update` -> `render` 60/120 times a second, so it doesn't need to re-execute anything when a provider updates, so you could pretty much use `ref.read` directly in `update` and it would work, but I suspect that `ref.listen` becomes cleaner and a bit more efficient?

Reply:
That's not about the 60/120 times a second. It's about having the rendering logic be more declarative and independent from when the component is mounted.

Using ref.listen you might need a `fireImmediately: true` for instance, and it's very imperative.


There's also the concern of keeping autoDispose providers alive if they are used (which ref.read wouldn't handle).

Reply:
Do you see a use case for interacting with providers within a Flame game component when it is not considered "alive"?

Reply:
I do not understand that question sorry

Reply:
Not sure we're on the same page about the component lifecycle so I want to ensure there's a shared understanding of a key concept:

Components are mounted once, and are removed / unmounted at some point during a game (e.g. enemy dies) - in this way they are distinct from Widgets that we understand as being unmounted / re-mounted as part of a re-render

I don't understand the case for decoupling the interactions with providers from when a component is mounted

Reply:
I don't think there's a real difference here. You could most definitely make a game engine controlled by widgets. Some people have done this in React.  
Currently Flame components seems more similar to a RenderObject than a Widget though. But it's likely possible to use a widget abstraction to manipulate them.

That's out of the scope of this PR of course. But Riverpod part of the point of Riverpod is to have business logic be more "widget-like" instead of the old imperative syntax.   
You're not meant to use `ref.listen` for too much.